### DISCUSSION TITLE
Deep nested object with useFieldArray and watched values.

### DISCUSSION THREAD

**Discussion by lostpfg** [react-hook-form/react-hook-form#3323] [Q&A]
Hello all,
Recently i started using the react-hook-from in my app and i have changed most of the forms to work the react-hook-form, but i have a case that i cannot manage to make any progress and it seems that there are some mistakes i cannot manage to solve.

My form is based on a nested object which has the following structure. _(Typescript definition with a simplified version of the actual implementation)_

```
type model = {
   tag: string;
   visible: boolean;
   columns?: (modelColumn | model)[];
}
type modelColumn = {
   property: string;
}
```

As you can see a model object reference itself at a deep nested level. The logic above that is that i want describe a model and i want to have the ability to switch a column with type `modelColumn` to `model` and have a nested object again in it's place.

Here you can see a sample data 
```
const initialData = {
  datasource: {
    tag: "tag1",
    visible: true,
    columns: [
      {
        property: "property"
      },
      {
        property: "property1"
      },
      {
        tag: "tag2",
        visible: true,
        columns: [
          {
            property: "property"
          },
          {
            tag: "tag3",
            visible: false,
            columns: [
              {
                property: "property"
              }
            ]
          }
        ]
      },
      {
        tag: "tag4",
        visible: false
      }
    ]
  }
};
```

After all i have created my logic with a Form Provider that has two main components that each on calls another in order to have the  nested logic with the hook form and the usage of `useFieldArray`.

here is a sandbox link of my implementation 
[https://codesandbox.io/s/hopeful-wilbur-qy8kj](https://codesandbox.io/s/hopeful-wilbur-qy8kj)
with the form that loads correctly the initial data of the form but with several issues after that.
At any case the `remove` method of the `useFieldArray` remove the last index of it's reference array, also also removing a column meshes with the nested values and their data.

So even with a simple logic of removing only values of type `modelColumn` structure of the object is always damaged after remove is called.

Additionaly `setValue` doesn't seem to work at all and the switch form `modelColumn` to `model` has unexpected outcome.

 Is there any advice on how to handle such a nested form with the usage of the provided hooks and methods with having the expected outcome.

Thanks in advance

**Comment by bluebill1049** [comment]
<img width="505" alt="Screen Shot 2020-11-02 at 6 02 41 pm" src="https://user-images.githubusercontent.com/10513364/97839096-9bd79080-1d35-11eb-81eb-76a16c78dbfb.png">

i think your problem is having a consistent data shape for your useFieldArray, what you having able is not going to work. there is no input being register with hook form, this is going to be failed with any of the actions that useFieldArray provided.

**Reply by bluebill1049** [reply]
so task 1: get your field Array with a consistent data structure 

```
  datasource: {
    tag: "tag1",
    visible: true,
    columns: [
      {
        property: "property" // missing other property
      },
      {
        property: "property1" // missing other property
      },
      {
        tag: "tag2", // this object probably should be your object shape
        visible: true,
        columns: [
          {
            property: "property"
          },
          {
            tag: "tag3",
            visible: false,
            columns: [
              {
                property: "property"
              }
            ]
          }
        ]
      },
      {
        tag: "tag4", // missing other property
        visible: false
      }
    ]
  }
};
```
